### PR TITLE
feat(dms/kafka): support to change the storage capacity and bandwidth of Kafka instance

### LINKED ISSUE #1774: [DMS] I want to change the storage capacity and bandwidth of Kafka instance
I want to change the storage capacity and bandwidth of Kafka instance

### PR DESCRIPTION
**Pull request by chengxiangdong**
**What this PR does / why we need it**:

- Support to resize the storage capacity  of the Kafka instances.
- Deprecated `available_zones`, use `availability_zones` instead, AZ code is easier to use.
- Deprecated  `bandwidth`, this value is not used in the service.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1774

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

### huaweicloud_dms_kafka_instance
```
make testacc TEST='./huaweicloud/services/acceptance/dms' TESTARGS='-run TestAccDmsKafkaInstances'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms -v -run TestAccDmsKafkaInstances -timeout 360m -parallel 4
=== RUN   TestAccDmsKafkaInstances_basic
=== PAUSE TestAccDmsKafkaInstances_basic
=== RUN   TestAccDmsKafkaInstances_withEpsId
=== PAUSE TestAccDmsKafkaInstances_withEpsId
=== CONT  TestAccDmsKafkaInstances_basic
=== CONT  TestAccDmsKafkaInstances_withEpsId
--- PASS: TestAccDmsKafkaInstances_withEpsId (599.39s)
--- PASS: TestAccDmsKafkaInstances_basic (945.98s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       946.054s

```

### huaweicloud_dms_product
```
make testacc TEST='./huaweicloud/services/acceptance/dms' TESTARGS='-run TestAccDmsProductV1DataSource'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms -v -run TestAccDmsProductV1DataSource -timeout 360m -parallel 4
=== RUN   TestAccDmsProductV1DataSource_basic
=== PAUSE TestAccDmsProductV1DataSource_basic
=== RUN   TestAccDmsProductV1DataSource_rabbitmqSingle
=== PAUSE TestAccDmsProductV1DataSource_rabbitmqSingle
=== RUN   TestAccDmsProductV1DataSource_rabbitmqCluster
=== PAUSE TestAccDmsProductV1DataSource_rabbitmqCluster
=== CONT  TestAccDmsProductV1DataSource_basic
=== CONT  TestAccDmsProductV1DataSource_rabbitmqCluster
=== CONT  TestAccDmsProductV1DataSource_rabbitmqSingle
--- PASS: TestAccDmsProductV1DataSource_rabbitmqSingle (52.33s)
--- PASS: TestAccDmsProductV1DataSource_basic (53.40s)
--- PASS: TestAccDmsProductV1DataSource_rabbitmqCluster (53.46s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       53.545s
```

### CODE REVIEW

File: docs/resources/dms_kafka_instance.md [hunk 1]
```diff
@@ -64,31 +54,22 @@ The following arguments are supported:
...
-
 * `storage_spec_code` - (Required, String, ForceNew) Specifies the storage I/O specification. Value range:
   + When bandwidth is 100MB: dms.physical.storage.high or dms.physical.storage.ultra
   + When bandwidth is 300MB: dms.physical.storage.high or dms.physical.storage.ultra
   + When bandwidth is 600MB: dms.physical.storage.ultra
-  + When bandwidth is 1200MB: dms.physical.storage.ultra
-
-  Changing this creates a new instance resource.
+  + When bandwidth is 1,200MB: dms.physical.storage.ultra
+    Changing this creates a new instance resource.
```
Comment (Lance52259) [inline docs/resources/dms_kafka_instance.md]: Deleting this blank line will cause formatting error.

File: docs/resources/dms_kafka_instance.md [hunk 2]
```diff
@@ -64,31 +54,22 @@ The following arguments are supported:
...
   consists of 4 to 64 characters, and supports only letters, digits, hyphens (-) and underscores (_).
 
-* `description` - (Optional, String) Specifies the description of the DMS kafka instance. It is a character string
-  containing not more than 1024 characters.
+* `product_id` - (Required, String) Specifies a product ID, which includes bandwidth, partition, broker and default
+  storage capacity.
+
+-> **NOTE:** Change this to change the bandwidth, partition and broker of the Kafka instances. Please note that the
+broker changes may cause storage capacity changes. So, if you specify the value of `storage_space`, you need to manually
+modify the value of `storage_space` after changing the `product_id`.
```
Comment (Lance52259) [inline docs/resources/dms_kafka_instance.md:62]: The parameter note need to be indented before.

File: docs/resources/dms_kafka_instance.md [hunk 3]
```diff
@@ -111,6 +90,13 @@ The following arguments are supported:
   the following character types: lowercase letters, uppercase letters, digits, and special characters (`~!@#$%^&*()-_
   =+\\|[{}]:'",<.>/?). Changing this creates a new instance resource.
 
+* `storage_space` - (Optional, Int) Specifies the message storage capacity, the unit is GB. Value range:
+  + When bandwidth is 100MB: 600–90000 GB
+  + When bandwidth is 300MB: 1200–90000 GB
+  + When bandwidth is 600MB: 2400–90000 GB
+  + When bandwidth is 1200MB: 4800–90000 GB
+    The storage capacity of the product used by default.
```
Comment (Lance52259) [inline docs/resources/dms_kafka_instance.md]: ditto

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go [hunk 4]
```diff
@@ -441,18 +520,58 @@ func resourceDmsKafkaInstanceUpdate(ctx context.Context, d *schema.ResourceData,
...
+func resizeInstance(ctx context.Context, d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	dmsV2Client, err := config.DmsV2Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.Errorf("error creating HuaweiCloud DMS instance client: %s", err)
+	}
+
+	product, err := getProductDetail(config, d)
+	if err != nil || product == nil {
+		return fmtp.Errorf("change storage_space failed, error querying product detail: %s", err)
```
Comment (Lance52259) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go:550]: If the error return of function `getProfuctDetail` is null and the product is empty, the error is reported as `change storage_space failed, error querying product detail: <nil>`, which is confusing

Comment (chengxiangdong) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go:550]: see `getProfuctDetail`, this case will not happen

Comment (Lance52259) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go:550]: does the `product == nil` needed?

File: docs/resources/dms_kafka_instance.md [hunk 5]
```diff
@@ -64,29 +54,21 @@ The following arguments are supported:
...
-* `storage_space` - (Required, Int, ForceNew) Specifies the message storage space. Value range:
-  + When bandwidth is 100MB: 600–90000 GB
-  + When bandwidth is 300MB: 1200–90000 GB
-  + When bandwidth is 600MB: 2400–90000 GB
-  + When bandwidth is 1200MB: 4800–90000 GB
-
-  Changing this creates a new instance resource.
-
 * `storage_spec_code` - (Required, String, ForceNew) Specifies the storage I/O specification. Value range:
   + When bandwidth is 100MB: dms.physical.storage.high or dms.physical.storage.ultra
```
Comment (hmbig2) [inline docs/resources/dms_kafka_instance.md:68]: maybe  put the explanation after spec_code  is better

File: docs/resources/dms_kafka_instance.md [hunk 6]
```diff
@@ -111,6 +91,14 @@ The following arguments are supported:
   the following character types: lowercase letters, uppercase letters, digits, and special characters (`~!@#$%^&*()-_
   =+\\|[{}]:'",<.>/?). Changing this creates a new instance resource.
 
+* `storage_space` - (Optional, Int) Specifies the message storage capacity, the unit is GB. Value range:
+  + When bandwidth is 100MB: 600–90000 GB
+  + When bandwidth is 300MB: 1200–90000 GB
+  + When bandwidth is 600MB: 2400–90000 GB
+  + When bandwidth is 1200MB: 4800–90000 GB
+
```
Comment (hmbig2) [inline docs/resources/dms_kafka_instance.md:99]: ditto

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_instance_test.go [hunk 7]
```diff
@@ -104,25 +103,35 @@ func TestAccDmsKafkaInstances_withEpsId(t *testing.T) {
...
 
-data "huaweicloud_dms_product" "test" {
-  engine        = "kafka"
-  instance_type = "cluster"
-  version       = "2.3.0"
+data "huaweicloud_dms_product" "test1" {
+  engine            = "kafka"
+  instance_type     = "cluster"
+  version           = "2.3.0"
+  bandwidth         = "100MB"
```
Comment (hmbig2) [inline huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_kafka_instance_test.go:125]: pls remove Deprecated arguments

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go [hunk 8]
```diff
@@ -307,100 +359,124 @@ func resourceDmsKafkaInstanceCreate(ctx context.Context, d *schema.ResourceData,
...
 	}
 	_, err = stateConf.WaitForStateContext(ctx)
 	if err != nil {
 		return fmtp.DiagErrorf("error waiting for instance (%s) to become ready: %s", v.InstanceID, err)
 	}
 
-	// Store the instance ID now
-	d.SetId(v.InstanceID)
+	// resize storage capacity of the instance
+	storage_space, ok := d.GetOk("storage_space")
```
Comment (niuzhenguo) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go]: better to check whether storage_space less than default storage_space before creating the instance.

**Pull request merged**